src/iamserver/streamwriter.hpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef STREAMWRITER_HPP_
+#define STREAMWRITER_HPP_
+
+#include <iamanager/v5/iamanager.grpc.pb.h>
+
+/**
+ * Controls writes to streams.
+ */
+template <typename T>
+class StreamWriter {
+public:
+    /**
+     * Closes all streams.
+     */
+    void Start()
+    {
+        std::unique_lock lock {mMutex};
+
+        mIsRunning      = true;
+        mNotificationID = 0;
+    }
+
+    /**
+     * Closes all streams.
+     */
+    void Close()
+    {
+        {
+            std::unique_lock lock {mMutex};
+
+            mIsRunning = false;
+            mLastMessage.reset();
+        }
+
+        mCV.notify_all();
+    }
+
+    /**
+     * Writes notification message to all streams.
+     *
+     * @param message notification message.
+     */
+    void WriteToStreams(const T& message)
+    {
+        {
+            std::unique_lock lock {mMutex};
+
+            ++mNotificationID;
+            mLastMessage = message;
+        }
+
+        mCV.notify_all();
+    }
+
+    /**
+     * Handles stream. Blocks the caller until the stream is closed.
+     *
+     * @param context server context.
+     * @param writer server writer.
+     * @return grpc::Status.
+     */
+    grpc::Status HandleStream(grpc::ServerContext* context, grpc::ServerWriter<T>* writer)
+    {
+        uint32_t lastNotificationID = 0;
+
+        while (mIsRunning && !context->IsCancelled()) {
+            std::shared_lock lock {mMutex};
+
+            bool res = mCV.wait_for(lock, cWaitTimeout, [this, lastNotificationID] {
+                return (mNotificationID != lastNotificationID && mLastMessage.has_value()) || !mIsRunning;
+            });
+
+            if (!mIsRunning) {
+                break;
+            }
+
+            if (res) {
+                // got notification, send it to the client
+                if (!writer->Write(*mLastMessage)) {
+                    break;
+                }
+
+                lastNotificationID = mNotificationID;
+            }
+        }
+
+        return grpc::Status::OK;
+    }
+
+private:
+    static constexpr auto cWaitTimeout = std::chrono::seconds(10);
+
+    std::atomic_bool            mIsRunning = true;
+    std::condition_variable_any mCV;
+    std::shared_mutex           mMutex;
```

Comment:
Why do you need it shared?

Reply:
And so what? If you don't use shared lock in any place then no need for shared mutex. Ordinary mutex matches this implementation.